src/inc/msquic.h

```diff
@@ -293,6 +293,82 @@ typedef struct QUIC_EXECUTION_CONFIG {
 #define QUIC_EXECUTION_CONFIG_MIN_SIZE \
     (uint32_t)FIELD_OFFSET(QUIC_EXECUTION_CONFIG, ProcessorList)
 
+//
+// Execution Context abstraction, which allows the application layer to
+// completely control execution of all MsQuic work.
+//
+
+typedef struct QUIC_EXECUTION_CONTEXT_CONFIG {
+    uint32_t IdealProcessor;
+    uint32_t PollingIdleTimeoutUs;
+    QUIC_EVENTQ* EventQ;
+} QUIC_EXECUTION_CONTEXT_CONFIG;
+
+typedef struct QUIC_EXECUTION_CONTEXT QUIC_EXECUTION_CONTEXT;
+
+//
+// This is called create the execution contexts.
+//
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+(QUIC_API * QUIC_EXECUTION_CREATE_FN)(
+    _In_ QUIC_EXECUTION_CONFIG_FLAGS Flags, // Used for datapath type
+    _In_ uint32_t Count,
+    _In_reads_(Count) QUIC_EXECUTION_CONTEXT_CONFIG* Configs,
+    _Out_writes_(Count) QUIC_EXECUTION_CONTEXT** ExecutionContexts
+    );
+
+//
+// This is called to delete the execution contexts.
+//
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+(QUIC_API * QUIC_EXECUTION_DELETE_FN)(
+    _In_ uint32_t Count,
+    _In_reads_(Count) QUIC_EXECUTION_CONTEXT** ExecutionContexts
+    );
+
+//
+// This is called to allow MsQuic to process any polling work. It returns the
+// number of milliseconds until the next scheduled timer expiration.
+//
+// TODO: Should it return an indication for if we should yield?
+//
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+uint32_t
+(QUIC_API * QUIC_EXECUTION_POLL_FN)(
+    _In_ QUIC_EXECUTION_CONTEXT* ExecutionContext
+    );
+
+//
+// This is called to allow MsQuic to process any completions that have occurred.
+//
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+uint32_t
+(QUIC_API * QUIC_EXECUTION_PROCESS_CQE_FN)(
+    _In_ QUIC_EXECUTION_CONTEXT* ExecutionContext,
+    _In_reads_(CqeCount) QUIC_CQE* Cqes,
```

Comment:
Need a function/way for the app to determine a CQE is for QUIC and not themselves.

Reply:
I added a function above, but it still has these assumptions.

Reply:
This needs to scale to work with arbitrary numbers of layers and unrelated libraries sharing an event queue. I don't think integer/bitfields will be sufficient, unless we want to operate a bitfield registry or perform dynamic bit allocations.

Reply:
With IOCPs on Windows, the standard way to solve this is to define an encapsulating struct containing an OVERLAPPED and whatever other data is required to provide app context per IO, e.g. an arbitrary function pointer and function pointer context. IDK if that will also be viable with epoll-style models, but I'd assume so.

Reply:
I don't think the struct itself needs to be consistent across platforms - you just need to define CxPlat wrappers for each operation. The suggestion above (function pointer, function pointer context, per IO) is sufficiently abstract and unambiguous, and should, I think, be viable on any platform.

Reply:
You also don't _really_ need to store a function pointer context itself, if you allow this CXPLAT_IO struct to be further encapsulated by callers, and simply pass the CXPLAT_IO during completion. Then they can unpack their own function pointer context, if they wish.

Reply:
Windows:

```
typedef struct CXPLAT_IO {
    OVERLAPPED Ov;
    VOID (*Completion)(struct CXPLAT_IO *Io);
};
```

Reply:
Linux:

struct CXPLAT_IO {
    CXPLAT_LISTENTRY Link; // IDK ?
    VOID (*Completion)(struct CXPLAT_IO *Io);
};

Reply:
And I suppose none of this should technically be restricted to "IO" and instead consider all of the above to mean "Work".